src/hotspot/share/services/memoryManager.cpp

```diff
@@ -114,34 +123,35 @@ instanceOop MemoryManager::get_memory_manager_instance(TRAPS) {
                            &args,
                            CHECK_NULL);
 
-    instanceOop m = (instanceOop) result.get_oop();
-    instanceHandle mgr(THREAD, m);
-
-    {
-      // Get lock before setting _memory_mgr_obj
-      // since another thread may have created the instance
-      MutexLocker ml(THREAD, Management_lock);
-
-      // Check if another thread has created the management object.  We reload
-      // _memory_mgr_obj here because some other thread may have initialized
-      // it while we were executing the code before the lock.
-      mgr_obj = Atomic::load(&_memory_mgr_obj).resolve();
-      if (mgr_obj != nullptr) {
-         return (instanceOop)mgr_obj;
-      }
-
-      // Get the address of the object we created via call_special.
-      mgr_obj = mgr();
-
-      // Use store barrier to make sure the memory accesses associated
-      // with creating the management object are visible before publishing
-      // its address.  The unlock will publish the store to _memory_mgr_obj
-      // because it does a release first.
-      Atomic::release_store(&_memory_mgr_obj, OopHandle(Universe::vm_global(), mgr_obj));
+    // Verify we didn't get a null manager.  If that could happen then we'd
+    // need to return immediately rather than continuing on and recording the
+    // manager has been created.
+    oop m = result.get_oop();
+    guarantee(m != nullptr, "Manager creation returned null");
+    instanceHandle mgr(THREAD, (instanceOop)m);
+
+    // Allocate global handle outside lock, to avoid any lock nesting issues
+    // with the Management_lock.
+    OopHandle mgr_handle(Universe::vm_global(), mgr());
+
+    // Get lock since another thread may have created and installed the instance.
+    MutexLocker ml(THREAD, Management_lock);
+
+    if (Atomic::load(&_memory_mgr_obj_initialized)) {
+      // Some other thread won the race.  Release the handle we allocated and
+      // use the other one.  Relaxed load is sufficient because flag update is
+      // under the lock.
+      mgr_handle.release(Universe::vm_global());
+    } else {
+      // Record the object we created via call_special.
+      _memory_mgr_obj = mgr_handle;
```

Comment:
Could assert that `_memory_mgr_obj.is_empty()`.

Reply:
Done.